### PR TITLE
Send and Receive

### PR DESCRIPTION
**Pull request by jnicklas**
This allows forks to send messages to each other and to receive them.

### CODE REVIEW

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-b3114fa](https://www.npmjs.com/package/effection/v/0.4.0-b3114fa) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 033fcac0259fc9b75791770befd5195dd0986e5d
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-da38c6e](https://www.npmjs.com/package/effection/v/0.4.0-da38c6e) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b2cfd261b7d7e4cb43a336150547fddf93b9b9a3
</p>

**Comment by jnicklas** [comment]
Tests for types currently fails on CI, but passes when I run them locally, maybe a difference in tsc version?

File: index.d.ts [hunk 1]
```diff
@@ -1,4 +1,10 @@
 declare module "effection" {
+  type PredicateFn = (value: any) => boolean;
+  type AnyTypeName = "undefined" | "object" | "boolean" | "number" | "bigint" | "string" | "symbol" | "function" | "array"
+  type PatternMatchObject = { [key: string]: PatternMatchOptions };
+  type Primitive = string | boolean | number | bigint | symbol | null | undefined;
```
Comment (jnicklas) [inline index.d.ts]: Seems like there *should* be a built in type for this? Couldn't find it though.

File: index.d.ts [hunk 2]
```diff
@@ -10,9 +16,18 @@ declare module "effection" {
     resume(result: T): void;
     throw(error: Error): void;
     halt(reason?: any): void;
+    send(message: any): void;
```
Comment (jnicklas) [inline index.d.ts]: We could limit the message to JSON-like objects, not sure if sending complex objects should be allowed. If we go down the path of messages being transported across different protocols, then sending complex objects should not be allowed.

File: src/pattern-match.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      return match.every((value, index) => patternMatch(value)(target[index]));
+    } else if(typeof(match) === "object") {
+      return Object.entries(match).every(([key, value]) => patternMatch(value)(target[key]));
+    } else if(typeof(match) === "function") {
+      return match(target);
+    } else {
+      return match === target;
+    }
+  };
+}
```
Comment (jnicklas) [inline src/pattern-match.js]: I made this a curried function, for ... no particular reason. Should it be?

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-93c2e36](https://www.npmjs.com/package/effection/v/0.4.0-93c2e36) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7e05ccb77ceb1734f4e4e756e2666807773b1c05
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-beda95a](https://www.npmjs.com/package/effection/v/0.4.0-beda95a) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d4c2bfd7076a25a3b05626288c4f0b1514d443cf
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-1c60ace](https://www.npmjs.com/package/effection/v/0.4.0-1c60ace) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b2cfd261b7d7e4cb43a336150547fddf93b9b9a3
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-b12aa2f](https://www.npmjs.com/package/effection/v/0.4.0-b12aa2f) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d8a96ef6014395cf3765076016aa2a867e6d3d1b
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-47c2735](https://www.npmjs.com/package/effection/v/0.4.0-47c2735) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c8bf945d15c5a63a36a9e275521052531ed496e1
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-7a355f6](https://www.npmjs.com/package/effection/v/0.4.0-7a355f6) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8eb3355f9719dfd675a90bb6330e77a745820bde
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `send-and-receive`.
You can try it out by running the following command:
```bash
$ npm install effection@send-and-receive
```
or by updating your package.json to:
```bash
{
  "effection": "send-and-receive"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.4.0-328a7a2](https://www.npmjs.com/package/effection/v/0.4.0-328a7a2) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 3a0ac7077c1488d3c4d5e3cff88649f72e6c9fbe
</p>

File: src/pattern-match.js (hunk 3 above)
Comment (cowboyd) [reply src/pattern-match.js]: It feels right. It seems like this is simpler to "compiling" a regex and then using it repeatedly to match. Right now, all of the matching happens dynamically, and so the only overhead that we're saving on is the actual allocation of the `matcher` function, but I can imagine a world where the optimizations are significant enough that they warrant pre-allocation of some memory in order to save some computational cycles as they are matched repeatedly.

If we make the API support this right out of the gate, then there's no need to change it going back.

Fun tip: A nice way that I've discovered to make curried APIs in JavaScript is with the `=>` syntax:

```js
const patternMatch = match => target => {
  if (match === undefined) {
  }
  //...
}
```

Of course, if you need to do some work in between arguments, for the pre-allocation, then it doesn't help, but for everything else....

**Comment by cowboyd** [comment]
superseded by #67

**Pull request closed without merging**